### PR TITLE
3060 active user session

### LINKED ISSUE #3060: As a TDP user, I need to stay logged in when I'm actively using the system
**Description:**

After signing in via acf ams or login.gov, and actively using the system, I am automatically signed out after 30 minutes. This is disruptive, and I'd like my session to persist unless:

1. I close my browser
2. I am inactive for 15-30 minutes (whichever is consistent with #3020)

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] A user session persists while the user is actively using the system
- [ ] the session ends when the browser is closed
- [x] the session ends if the user is inactive for 15-30 minutes (see above)
- [x] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] update session management-related code docs and markdowns
- [x] ensure consistency between sessions via login.gov vs acf ams
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- _Note 1_
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- #3020 

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes

Pull request closes #3060 _

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. The timeout is set 10 seconds, as long as you browse different pages, you will be considered active.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] A user session persists while the user is actively using the system
- [ ] ~~the session ends when the browser is closed~~
- [x] the session ends if the user is inactive for 15-30 minutes (see above)
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by raftmsohani** [comment]
backend is using ```django.contrib.session```, which enables db backend session manager. However, we also enable cookie session enabled by ```SESSION_ENGINE = "django.contrib.sessions.backends.signed_cookies"``` which in turn happens to only respect default ```SESSION_COOKIE_AGE``` and not updated session timeout

**Comment by raftmsohani** [comment]
We are using [signed-cookies](https://github.com/django/django/blob/39de2e97a06d0317973b280bc159ca6f89fc64e3/django/contrib/sessions/backends/signed_cookies.py#L17) to keep the session information, which is using [_SessionBase_](https://github.com/django/django/blob/39de2e97a06d0317973b280bc159ca6f89fc64e3/django/contrib/sessions/backends/base.py#L265). 
With every new request, the session timestamp is updated by signing a new session using [signing](https://github.com/django/django/blob/39de2e97a06d0317973b280bc159ca6f89fc64e3/django/core/signing.py#L274) package.

File: tdrs-backend/tdpservice/core/custom_session_engine.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        if signature fails.
+        """
+
+
+        try:
+            return signing.loads(
+                self.session_key,
+                serializer=self.serializer,
+                # This doesn't handle non-default expiry dates, see #19201
+                max_age=datetime.timedelta(days=1), #self.get_session_cookie_age(), #
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/core/custom_session_engine.py]: we are only overriding max_age. By default `signed_cookies` assigns default `SESSION_COOKIE_AGE`, which will expire the session. We cannot use the default session `set_expiry` for signed sessions, unless we unsign and re-sign the session with every request which is more expensive.

File: tdrs-backend/tdpservice/settings/common.py [hunk 2]
```diff
@@ -281,10 +281,12 @@ class Common(Configuration):
...
     # Sessions
-    SESSION_ENGINE = "django.contrib.sessions.backends.signed_cookies"
+    #SESSION_ENGINE = "django.contrib.sessions.backends.signed_cookies"
+    SESSION_ENGINE = "tdpservice.core.custom_session_engine"
     SESSION_COOKIE_HTTPONLY = True
-    SESSION_EXPIRE_AT_BROWSER_CLOSE = True
-    SESSION_COOKIE_AGE = 15 * 60  # 15 minutes
+    #SESSION_EXPIRE_AT_BROWSER_CLOSE = True
+    SESSION_SAVE_EVERY_REQUEST = True
+    SESSION_COOKIE_AGE = 10  # 15 minutes
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/settings/common.py]: For testing only: revert back to previous setting

File: tdrs-backend/tdpservice/data_files/admin/admin.py [hunk 3]
```diff
@@ -1,7 +1,6 @@
 """Admin class for DataFile objects."""
 from django.contrib import admin
 from tdpservice.core.utils import ReadOnlyAdminMixin
-# from tdpservice.core.filters import custom_filter_title
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/data_files/admin/admin.py:4]: Unrelated to this PR but I saw this is commented and not deleted

File: tdrs-backend/docs/session-management.md [hunk 4]
```diff
@@ -11,7 +11,7 @@ When the user logs in, they will receive an HttpOnly cookie with no `Expires=` s
 SESSION_EXPIRE_AT_BROWSER_CLOSE=True
 ```
 
-The cookie itself contains a `sessionid` reference to a Django-managed session. The session expiration is set to the same expiration of the login.gov-provided jwt, **15 minutes**.
+The cookie itself contains a `sessionid` reference to a Django-managed session. By deafult, the session expiration is set to the same expiration of the login.gov-provided jwt, **15 minutes**. Since `sessionid` is being signed when created, it not possible to update the expiry without decoding and recreating the signature, so the session expires even though the cookie is being extended with every request. Thus, in order to overcome this shortcoming, a longer expiry is being assigned to `sessionid` using a new variable `SIGNED_COOKIE_EXPIRES` in common settings.
```
Comment (andrew-jameson) [inline tdrs-backend/docs/session-management.md:14]: Documentation updates! ❤️

**Comment by elipe17** [comment]
@raftmsohani I am trying to test this and I am not seeing the expected behavior. I have cleared my cookies, logged in, and waited for longer than 10 seconds but still remain logged in. Is there something I am missing?

**Comment by raftmsohani** [comment]
> @raftmsohani I am trying to test this and I am not seeing the expected behavior. I have cleared my cookies, logged in, and waited for longer than 10 seconds but still remain logged in. Is there something I am missing?

No it should work, I will double check and get back

File: tdrs-backend/tdpservice/settings/common.py [hunk 5]
```diff
@@ -281,10 +281,12 @@ class Common(Configuration):
     )
 
     # Sessions
-    SESSION_ENGINE = "django.contrib.sessions.backends.signed_cookies"
+    SESSION_ENGINE = "tdpservice.core.custom_session_engine"
+    SIGNED_COOKIE_EXPIRES = 60 * 60 * 24  # 24 hours
     SESSION_COOKIE_HTTPONLY = True
-    SESSION_EXPIRE_AT_BROWSER_CLOSE = True
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/settings/common.py:286]: The setting causes the session expiry to be set to browser close and ignore SESSION_COOKIE_AGE.

Comment (raftmsohani) [reply tdrs-backend/tdpservice/settings/common.py:286]: We cannot set expire the session at browser close together with session expiry age.

File: tdrs-backend/tdpservice/core/custom_session_engine.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                # This doesn't handle non-default expiry dates, see #19201
+                max_age=datetime.timedelta(seconds=settings.SIGNED_COOKIE_EXPIRES),
+                salt="django.contrib.sessions.backends.signed_cookies",
+            )
+        except Exception:
+            # BadSignature, ValueError, or unpickling exceptions. If any of
+            # these happen, reset the session.
+            return {}
+
+    def cycle_key(self):
```
Comment (elipe17) [inline tdrs-backend/tdpservice/core/custom_session_engine.py]: Can we remove these now that we don't need to debug anymore?

**Review by elipe17 — APPROVED** [review]
Not sure why it wasn't working initially. But it worked like a charm this morning!

**Comment by jtimpe** [comment]
> @raftmsohani I am trying to test this and I am not seeing the expected behavior. I have cleared my cookies, logged in, and waited for longer than 10 seconds but still remain logged in. Is there something I am missing?

i'm still experiencing this today - was there anything you did differently @elipe17?

**Comment by raftmsohani** [comment]
> > @raftmsohani I am trying to test this and I am not seeing the expected behavior. I have cleared my cookies, logged in, and waited for longer than 10 seconds but still remain logged in. Is there something I am missing?
> 
> i'm still experiencing this today - was there anything you did differently @elipe17?

@jtimpe if you pull the latest version, I have disabled `SESSION_EXPIRE_AT_BROWSER_CLOSE`, and you should see the correct behavior now

**Review by jtimpe — APPROVED** [review]

File: tdrs-backend/tdpservice/settings/common.py [hunk 7]
```diff
@@ -281,10 +281,12 @@ class Common(Configuration):
     )
 
     # Sessions
-    SESSION_ENGINE = "django.contrib.sessions.backends.signed_cookies"
+    SESSION_ENGINE = "tdpservice.core.custom_session_engine"
+    SIGNED_COOKIE_EXPIRES = 60 * 60 * 24  # 24 hours
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/settings/common.py]: @raftmsohani is this set to expire 24 hours from the initial session? if so, i wonder if it would be better to shorten this to 12 hours, which is on the very high end of an average work day. i'm also reaching out to our security officer about this.

Comment (ADPennington) [reply tdrs-backend/tdpservice/settings/common.py]: evidence below suggests expiry 24 hours into future. left side is staging (current state) and right side is qasp (this branch)

Screenshot 2024-10-10 172400

Comment (ADPennington) [reply tdrs-backend/tdpservice/settings/common.py]: @raftmsohani TDP's ISSO's feedback: 

- this change is related to [AC-2-5](https://github.com/raft-tech/TANF-app/blob/develop/docs/Security-Compliance/Security-Controls/ac-2-5/index.md). We'll need to update this documentation as part of this work to capture how the session refresh/expiration works for ACF users. 
- let's also reduce the expiration to 12 hours
- we'll likely need to also complete risk-based decision documentation when the security team provides it.

**Review by ADPennington — CHANGES_REQUESTED** [review]
feedback included in comments.

File: tdrs-backend/tdpservice/settings/common.py (hunk 7 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/settings/common.py]: @ADPennington we shouldn't need to update the AC-2-5, since we are still following same time out concept. In fact, this PR makes sure we are exactly following this time out setting.
The only difference is: with new changes, even if the user is active, system will log the user out after 12hrs

File: tdrs-backend/tdpservice/settings/common.py [hunk 8]
```diff
@@ -282,11 +282,10 @@ class Common(Configuration):
 
     # Sessions
     SESSION_ENGINE = "tdpservice.core.custom_session_engine"
-    SIGNED_COOKIE_EXPIRES = 60 * 60 * 24  # 24 hours
+    SIGNED_COOKIE_EXPIRES = 60 * 60 * 12  # 12 hours
     SESSION_COOKIE_HTTPONLY = True
     SESSION_SAVE_EVERY_REQUEST = True
-    #SESSION_EXPIRE_AT_BROWSER_CLOSE = True
-    SESSION_COOKIE_AGE = 10  # 15 minutes
+    SESSION_COOKIE_AGE = 60 * 30  # 30 minutes
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/settings/common.py:288]: @raftmsohani can confirm that I received the warning for inactivity while logged in via AMS. im gonna wait another 30 minutes, because Im trying to understand the relationship between cookie_age and cookie_expires. max age set to 30 minutes when looking at devtools.

File: tdrs-backend/tdpservice/settings/common.py (hunk 7 above)
Comment (ADPennington) [reply tdrs-backend/tdpservice/settings/common.py]: confirming that the AC-2-5 seems aligned with the behavior in this PR, particularly this part: `If the user chooses to extend the session, TDP will refresh the session.`

**Review by ADPennington — APPROVED** [review]
🚀 thanks @raftmsohani
